### PR TITLE
Speed up partest

### PR DESCRIPTION
**Pull request by retronym**
(2.12.x version of #6391)

Prepares for, and then switches to, partest 1.1.6 which will
include a mode to run tests without forking.

Removes some resource hungry tests. Volunteers sought to revive
these in more frugal form.

Also backports the new test listener that integrates partest
into the Jenkins Test Results report.

### CODE REVIEW

**Comment by retronym** [comment]
My vote is to merge this and merge forward to 2.13. If we find a problem with Partest 1.1.7, we can just bump the version back without the other changes causing problems. @lrytz WDYT?

File: src/partest-extras/scala/tools/partest/BytecodeTest.scala [hunk 1]
```diff
@@ -133,7 +133,7 @@ abstract class BytecodeTest {
     // logic inspired by scala.tools.util.PathResolver implementation
     // `Settings` is used to check YdisableFlatCpCaching in ZipArchiveFlatClassPath
     val factory = new ClassPathFactory(new Settings())
-    val containers = factory.classesInExpandedPath(Defaults.javaUserClassPath)
+    val containers = factory.classesInExpandedPath(sys.props("partest.output") + ":" + Defaults.javaUserClassPath)
```
Comment (lrytz) [inline src/partest-extras/scala/tools/partest/BytecodeTest.scala]: This should probably use `java.io.File.pathSeparator` to also work on windoze

File: src/partest-extras/scala/tools/partest/ReplTest.scala [hunk 2]
```diff
@@ -19,6 +19,13 @@ abstract class ReplTest extends DirectTest {
   final override def settings: Settings = {
     val s = super.settings
     s.Xnojline.value = true
+    if (getClass.getClassLoader.getParent != null) {
+      s.classpath.value = s.classpath.value match {
+        case "" => testOutput.toString
+        case s => s + ":" + testOutput.toString
```
Comment (lrytz) [inline src/partest-extras/scala/tools/partest/ReplTest.scala]: also `java.io.File.pathSeparator`

**Review by lrytz — CHANGES_REQUESTED** [review]
Agree to merge it here.
  - Why did you backport "Elide prefixes in printed types uniformly in runtime reflection"?
  - We could create a `test/longrunning` folder for the resource hungry tests, and run them in nightlies for example. WDYT?
  - Could you expand a little in the first commit message "Fork tests that inherently require it"? What are known conditions? For example, I randomly picked `t8266-octal-interp.scala` and was wondering why. Is it the mix of compiler an run output in the check file?

File: src/partest-extras/scala/tools/partest/ReplTest.scala (hunk 2 above)
Comment (retronym) [reply src/partest-extras/scala/tools/partest/ReplTest.scala]: pushed an extra commit

**Comment by retronym** [comment]
> Why did you backport "Elide prefixes in printed types uniformly in runtime reflection"?

Prior to that fix, tests that relied on `Type.toString` from `runtime.universe` would change depending on classloader details that differed between forked and unforked test execution.

> We could create a test/longrunning folder for the resource hungry tests, and run them in nightlies for example. WDYT?

I don't think there is enough value in the ones I've dropped to setup that infrastructure.

> Fork tests that inherently require it

Can do. Once I got down to the last few tests, I wasn't always sure why the forking was necessary, but for some I should be able to give an explanation.

**Comment by lrytz** [comment]
>> Fork tests that inherently require it
> 
> Can do. Once I got down to the last few tests, I wasn't always sure why the forking was necessary, but for some I should be able to give an explanation.

Just a comment here is fine, instead of rebasing and rebuilding the whole thing. It doesn't need to be exhausitive, just a few pointers.

**Comment by retronym** [comment]
Know reasons to fork:

 - testing `@native` (test/files/jvm/natives)
 - tests for our facade over JVM shutdown hooks

Less well known:
 - tests for memory leaks that are somewhat sensitive to the heap size and seemed more stable to partition into a separate VM
 - using less well known features of partest like including `methvsfield.jar` on the app classpath (`**/methvsfield.*`)

Not known/remembered:
  - ` test/files/run/t8266-octal-interp`

File: src/partest-extras/scala/tools/partest/ReplTest.scala (hunk 2 above)
Comment (lrytz) [reply src/partest-extras/scala/tools/partest/ReplTest.scala]: > pushed

don't see it yet 🙈

**Comment by retronym** [comment]
> don't see it yet 🙈

🐵

**Review by lrytz — APPROVED** [review]

**Comment by retronym** [comment]
Submitted the forward merge as https://github.com/scala/scala/pull/6451

**Pull request merged**